event-web-framework/draft.markdown

```diff
@@ -0,0 +1,872 @@
+The binding happens in the last line; `(setf (gethash ,uri *handlers*) ,handler)`, which is what hash-table assignments look like in Common Lisp (modulo the commas, which are part of our macro.) Note that the `assert` is outside of the quoted area, which means that it'll be run as soon as the macro is _called_ rather than when its result is evaluated.
+
+When we further expand our expansion of the `send-message` `define-handler` above, we get:
+
+```lisp
+    (PROGN
+     (WHEN (GETHASH "/send-message" *HANDLERS*)
+       (WARN "Redefining handler '/send-message'"))
+     (SETF (GETHASH "/send-message" *HANDLERS*)
+             (MAKE-CLOSING-HANDLER (:CONTENT-TYPE "text/html")
+                 ((ROOM :STRING (LEN-BETWEEN 0 ROOM 16))
+                  (NAME :STRING (LEN-BETWEEN 1 NAME 64))
+                  (MESSAGE :STRING (LEN-BETWEEN 5 MESSAGE 256)))
+               (PUBLISH! (INTERN ROOM :KEYWORD)
+                         (ENCODE-JSON-TO-STRING
+                          `((:NAME ,@NAME) (:MESSAGE ,@MESSAGE)))))))
+```
+
+This is starting to look more like a custom implementation of what we would have written to marshal a request from a URI to a handler function had we written it all ourselves. But we didn't have to!
+
+We still have `make-closing-handler` left to go in our expansion. Here is its definition:
+
+```lisp
+	(defmacro make-closing-handler ((&key (content-type "text/html")) (&rest args) &body body)
+	  `(lambda (sock parameters)
+	     (declare (ignorable parameters))
+	     ,(arguments args
+			 `(let ((res (make-instance 
+				      'response 
+				      :content-type ,content-type 
+				      :body (progn ,@body))))
+			    (write! res sock)
+			    (socket-close sock)))))
+```
+
+So making a closing-handler involves making a `lambda`, which is just what you call anonymous functions in Common Lisp. We also set up an interior scope that makes a `response` out of the `body` argument we're passing in, `write!`s that to the requesting socket, then closes it. The remaining question is, what is `arguments`?
+
+```lisp
+	(defun arguments (args body)
+	  (loop with res = body
+	     for arg in args
+	     do (match arg
+		  ((guard arg-sym (symbolp arg-sym))
+		   (setf res `(let ((,arg-sym ,(arg-exp arg-sym)))
+				,res)))
+		  ((list* arg-sym type restrictions)
+		   (setf res
+			 `(let ((,arg-sym ,(or (type-expression (arg-exp arg-sym) type restrictions) (arg-exp arg-sym))))
+			    ,@(awhen (type-assertion arg-sym type restrictions) `((assert-http ,it)))
+			    ,res))))
+	     finally (return res)))
+```
+
+Welcome to the hard part. `arguments` turns the validators we registered with our handler into a tree of parse attempts and assertions. `type-expression`, `arg-exp`, and `type-assertion` are used to implement and enforce a "type system" for the kinds of data we're expecting in our responses; we'll discuss them in SECTION FIXME. Using this together with `make-closing-handler`, this would the validation rules we wrote here:
```

Comment:
I assume that should read "Using this together with `make-closing-handler` would implement the validation rules we wrote here:"

Not sure the sentence makes sense anymore, given that this is the first place we introduce `define-handler` at this point. I think I'll need to finesse some of the prose based on these edits.